wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,393 @@
+
+-- | Informs the 'WalletSubmission' layer about new 'Pending' transactions.
+addPending :: WalletSubmission m -> Pending -> WalletSubmission m
+addPending ws newPending =
+    let ws' = ws & over (wsState . wssPendingSet) (unionPending newPending)
+    in schedule ws' newPending
+
+-- | Removes the input 'Pending' from the local 'WalletSubmission' pending set.
+remPending :: WalletSubmission m -> Pending -> WalletSubmission m
+remPending ws updatedPending =
+    ws & over (wsState . wssPendingSet) (flip differencePending updatedPending)
+
+-- | A \"tick\" of the scheduler.
+-- Returns the set transactions which needs to be droppped by the system as
+-- they likely exceeded the submission count and they have no chance to be
+-- adopted in a block, or that have simply been successfully retrasmitted.
+-- @N.B.@ The returned 'WalletSubmission' does not come with an already-pruned
+-- local 'Pending' set. It will be caller's responsibility to call 'remPending'
+-- on the 'WalletSubmission'.
+--
+tick :: Monad m
+     => WalletSubmission m
+     -- ^ The current 'WalletSubmission'.
+     -> m (Evicted, WalletSubmission m)
+     -- ^ The set of transactions upper layers will need to drop, together
+     -- with the new 'WalletSubmission'.
+tick ws = do
+    let wss               = ws  ^. wsState
+        currentSlot       = wss ^. wssCurrentSlot
+        scheduler         = ws  ^. getScheduler
+        rho               = _wsResubmissionFunction ws
+        dueNow            = dueThisSlot currentSlot ws
+    (toPrune, scheduler') <- rho currentSlot dueNow scheduler
+    let newState = ws & over (wsState . wssScheduler) (const (purgeSlot currentSlot scheduler'))
+                      . over (wsState . wssCurrentSlot) succ
+    return (toPrune , newState)
+
+-- | Returns a set of 'Pending' transactions which are due in the given
+-- 'Slot'. First of all, it looks at the 'Scheduled' items (transactions) due
+-- this slot, and sorts them topologically, and:
+--
+-- 1. If a topological order can be found, the 'Scheduled' items are returned.
+-- 2. TODO(adn) If a topological order can't be found, we will need to deal with it.
+--
+dueThisSlot :: Slot
+            -- ^ The current 'Slot'.
+            -> WalletSubmission m
+            -- ^ The 'WalletSubmissionState'.
+            -> [Scheduled]
+            -- ^ The filtered set of 'Pending' which are due in the input slot.
+dueThisSlot currentSlot ws =
+    let scheduler  = ws ^. getScheduler . to _ssScheduled
+        slotted    = fromMaybe mempty (IntMap.lookup currentSlot scheduler)
+        topSorted  = topsortTxs toTx slotted
+    in case topSorted of
+            Nothing     -> slotted
+            Just sorted -> sorted
+    where
+        toTx :: Scheduled -> WithHash Core.Tx
+        toTx (Scheduled (txId, txAux, _)) =  WithHash (Core.taTx txAux) txId
+
+-- | Overrides the 'Scheduler' with an input transaction and a
+-- custom 'Schedule'. Useful to force dispatching, especially in tests.
```

Comment:
Would be good to document what's unsafe about this. Is this used only for the tests? If so, might want to document that.